### ISSUE TITLE
Proper numeric tower

### ISSUE THREAD

**Issue by mattwparas** [mattwparas/steel#62]
It would be nice to have a proper numeric tower, right now there are integers that promote to floats at overflow - but otherwise bigints are not implemented - historically this was to avoid having to box all integers.

It might be worth implementing now, and just wrap up this crate https://docs.rs/num/latest/num/

**Comment by wmedrano** [comment]
What's the status of this? It seems like `BigInts` are implemented. The following seems just fine in the terminal:

```scheme
λ > (define x 10)
λ > (+ x 1000000000000000000000000000000000000000000000000000000000000000000000)
=> 1000000000000000000000000000000000000000000000000000000000000000000010
```

**Comment by mattwparas** [comment]
yeah, big ints are supported now (of course, probably missing a few things here and there) but rationals are not yet supported. That would be next on the list

**Comment by wmedrano** [comment]
For implementation, what do you think of adding another variant to `SteelVal`:

```rust
enum Fraction {
    Small(Rational32),
    Big(BigRational),
}

enum SteelVal {
    ...
    FractV(Fraction),
    ...
}

// We should stay under this constraint when adding a new type.
const _ASSERT_SMALL: () = assert!(std::mem::size_of::<SteelVal>() <= 16);
```

If it looks alright, I may try prototyping it.

**Comment by mattwparas** [comment]
This looks alright - unfortunately you might have to `Gc` it like how `BigNum` is:

```rust
BigNum(Gc<num::BigInt>),
```

the rational 32 might be able to be unboxed, however to do that you'd have to do lift the enum variant up I think unless some enum variant size optimization kicks in.

If this works, that is pleasant:

```rust
enum Fraction {
    Small(Rational32),
    Big(Gc<BigRational>),
}

enum SteelVal {
    ...
    FractV(Fraction),
    ...
}
```

However you might have to do this:

```rust
enum SteelVal {
    ...
    SmallFract(Rational32),
    BigFract(Gc<BigRational>),
    ...
}
```

**Comment by mattwparas** [comment]
Otherwise - definitely give prototyping it a shot

**Comment by wmedrano** [comment]
Is the goal to implement the full numerical tower? If so, do you have a checklist of requirements? From skimming [Wikipedia](https://en.wikipedia.org/wiki/Numerical_tower) it seems like `complex` numbers are the only thing left.

**Comment by mattwparas** [comment]
This could be a general reference: https://www.gnu.org/software/guile/manual/html_node/Numerical-Tower.html

Yes, the goal is to implement the full tower, and I think you're right, all that is missing is complex numbers.

**Comment by wmedrano** [comment]
Cool, so to recap for Steel.

| Type            | Representation                        |
|-----------------|---------------------------------------|
| integer         | `isize` or `BigInt`                   |
| rational number | `Rational<i32, i32>` or `BigRational` |
| real number     | `f64`                                 |
| complex         | undecided                             |

What do you think of using `Complex<f64, f64>` as the `complex` representation?

**Comment by mattwparas** [comment]
I think this seems reasonable. At first I would have naively assumed there is like some annoyingly complicated hierarchy of int -> big int -> real for complex numbers too, but that seems unnecessarily complex (no pun intended)

Also yeah that equation seems to imply the real number precision. I'm good with complex just being two floats. If a need for higher precision presents itself in the future for whatever reason we can cross that bridge when we get there. I don't foresee it being necessary

**Comment by mattwparas** [comment]
Well that is annoying isn't it, I think I'd be happy to go with whatever is the easiest to implement since I'm not the one doing it and also don't have a strong opinion

**Comment by wmedrano** [comment]
Since it's going to be boxed anyways, the simplest implementation will probably be a `Gc` that points to 2 non-complex numbers:

```rust
struct SteelComplex {
    re: SteelVal,
    im: SteelVal
}
```

I kind of like this. Though I confess I don't really have a use for complex numbers and neither does anyone in my small circle

**Comment by mattwparas** [comment]
Yeah I agree that seems like the best option, To play devils advocate I guess you could have like:

```rust
enum NumberKind {
   // whole numeric tower in here unboxed
}
```

But... that is just gnarly and this way you can inherit some of the existing functions when doing any promotion logic, and also since I don't particularly have a need for it (nor do you) then I don't think it needs to be hyper optimized right out of the gate. I also don't think it will make a particularly big difference

So tl;dr go for it

**Comment by wmedrano** [comment]
# Current Status
- exact-integer-sqrt
- floor/
- floor-quotient
- floor-remainder
- gcd
- lcm
- max
- min
- modulo
- rationalize
- remainder
- truncate
- truncate/
- truncate-quotient
- truncate-remainder
- nan?
- zero?
- positive?
- negative?
- sqrt
- make-rectangular
- make-polar
- real-part
- imag-part
- magnitude
- angle
- inexact
- exact
